### PR TITLE
Initial draft of an object graph manifest

### LINKED ISSUE #3536: Add a manifest for backup/restore
**Describe the problem/challenge you have**
We would like Velero to be able to:

- Have a dry-run option that allows users to see what Velero is going to do before it does a backup or restore
- Execute multiple backups or restores simultaneously
- Restart backups in the event of a server restart
- Order the restore of resources intelligently

**Describe the solution you'd like**
Currently, Velero does not have a complete manifest of everything in the backup, aside from the backup tarball itself.
This change introduces a new data structure to be stored with a backup in object storage which will allow for more efficient operations in reporting of what a backup contains.
Additionally, this manifest should enable advancements in Velero's features and architecture, enabling dry-run support, concurrent backup and restore operations, and reliable restoration of complex applications.

**Environment:**

- Velero version (use `velero version`):
- Kubernetes version (use `kubectl version`):
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "The project would be better with this feature added"
- :-1: for "This feature will not enhance the project in a meaningful way"

### PR DESCRIPTION
**Pull request by nrb**
Fixes #3536 

 # Please add a summary of your change

This change adds a design for adding a "manifest" data structure to Velero. Initially, this is just populating the structure for use, not actually relying on it for any of the stated enablement cases.

# Does your change fix a particular issue?

Does not fix a specific issue, but provides a basis for tackling #448, #2888, and more.

# Please indicate you've done the following:

[x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
[x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required`.
[x] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: design/graph-manifest.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+- API Group and Version (example: backup.velero.io/v1)
+- Namespace 
+- Name
+- Labels
+
+This criteria covers the majority of Velero's inclusion or exclusion logic.
+However, some additional fields enable further use cases.
+
+- Owners, which are other Kubernetes objects that have some relationship to this object. They may be strict or soft dependencies.
```
Comment (nrb) [inline design/graph-manifest.md:52]: There are objects that have relationships not called `owner`.

To properly handle these, I can think of 2 solutions

1) Use the definition (CR or Kubernetes in-tree) of the field to validate whether or not it's an object reference
2) Allow developers to supply an annotation on the CR describing the fields that are references.

I would prefer 1, but need to do a little research here.

File: design/graph-manifest.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+Here are some use cases that this data structure should enable, that have been difficult to implement prior to its existence:
+
+- A dry-run operation on backup, informing the user what would be selected if they were to perform the operation.
+ A manifest could be created and saved, allowing for a user to do a dry-run, then accept it to perform the backup.
+ Restore operations can be treated similarly.
+- Efficient, non-overlapping parallelization of backup and restore operations.
+ By building or reading a manifest before performing a backup or restore, Velero can determine if there are overlapping resources.
+ If there are no overlaps, the operations can proceed in parallel.
+ If there are overlaps, the operations can proveed serially.
```
Comment (alaypatel07) [inline design/graph-manifest.md:72]: NIT: s/proveed/proceed

**Review by dsu-igeek — APPROVED** [review]
Thanks, we will merge this as preliminary

**Review by zubron — APPROVED** [review]

**Pull request merged**